File: impl/ocaml/src/filters.ml
(* Filter and backdrop-filter utilities (SPEC §10.8, "Filters"). *)

open Ast
open Candidate
open Utilities

let property = Builtin_variants.property
let filter_names = [ "blur"; "brightness"; "contrast"; "grayscale"; "hue-rotate"; "invert"; "saturate"; "sepia"; "drop-shadow" ]
let backdrop_names = [ "blur"; "brightness"; "contrast"; "grayscale"; "hue-rotate"; "invert"; "opacity"; "saturate"; "sepia" ]
let composed_filter prefix names = String.concat " " (List.map (fun n -> "var(" ^ prefix ^ n ^ ",)") names)

(* The composed filter value. *)
let filter = composed_filter "--tw-" filter_names

(* The composed backdrop-filter value. *)
let backdrop_filter = composed_filter "--tw-backdrop-" backdrop_names

let filter_registrations backdrop =
  let prefix, names = if backdrop then ("--tw-backdrop-", backdrop_names) else ("--tw-", filter_names) in
  List.map (fun n -> property (prefix ^ n) None) names

let filter_declarations backdrop =
  if backdrop then [ decl "-webkit-backdrop-filter" backdrop_filter; decl "backdrop-filter" backdrop_filter ]
  else [ decl "filter" filter ]

(* Registers the filter and backdrop-filter utilities. *)
let register u theme =
  let stat_fn name f = static_utility_fn u name f in
  let fn root desc = functional_utility u theme root desc in
  let percentage = bare_suffix "%" and degrees = bare_suffix "deg" in
  List.iter
    (fun backdrop ->
      let prefix, variable = if backdrop then ("backdrop-", "--tw-backdrop-") else ("", "--tw-") in
      let keys name = if backdrop then [ "--backdrop-" ^ name; "--" ^ name ] else [ "--" ^ name ] in
      let emit name value = filter_registrations backdrop @ [ decl (variable ^ name) value ] @ filter_declarations backdrop in
      let direct value =
        if backdrop then [ decl "-webkit-backdrop-filter" value; decl "backdrop-filter" value ] else [ decl "filter" value ]
      in
      stat_fn (prefix ^ "filter-none") (fun () -> direct "none");
      stat_fn (prefix ^ "filter") (fun () -> filter_registrations backdrop @ filter_declarations backdrop);
      fn (prefix ^ "filter") (describe (fun value _ -> direct value));
      fn (prefix ^ "blur") (describe ~theme_keys:(keys "blur") (fun value _ -> emit "blur" ("blur(" ^ value ^ ")")));
      stat_fn (prefix ^ "blur-none") (fun () -> emit "blur" "");
      List.iter
        (fun name ->
          fn (prefix ^ name)
            (describe ~theme_keys:(keys name) ~handle_bare_value:percentage (fun value _ -> emit name (name ^ "(" ^ value ^ ")"))))
        [ "brightness"; "contrast"; "saturate" ];
      List.iter
        (fun name ->
          fn (prefix ^ name)
            (describe ~theme_keys:(keys name) ~default_value:"100%" ~handle_bare_value:percentage (fun value _ ->
                 emit name (name ^ "(" ^ value ^ ")"))))
        [ "grayscale"; "invert"; "sepia" ];
      fn (prefix ^ "hue-rotate")
        (describe ~theme_keys:(keys "hue-rotate") ~supports_negative:true ~handle_bare_value:degrees (fun value _ ->
             emit "hue-rotate" ("hue-rotate(" ^ value ^ ")")));
      if backdrop then
        fn "backdrop-opacity"
          (describe ~theme_keys:[ "--backdrop-opacity"; "--opacity" ] ~handle_bare_value:percentage (fun value _ ->
               emit "opacity" ("opacity(" ^ value ^ ")")))
      else begin
        stat_fn "drop-shadow-none" (fun () -> emit "drop-shadow" "");
        functional u "drop-shadow" (fun c ->
            let color value = handled [ property "--tw-drop-shadow-color" None; decl "--tw-drop-shadow-color" value ] in
            let wrap value =
              match Shadows.wrap_shadow_colors value c theme "--tw-drop-shadow-color" false with
              | None -> not_handled
              | Some replaced ->
                  let shadows = List.map (fun s -> "drop-shadow(" ^ String.trim s ^ ")") (Utils.segment replaced ',') in
                  handled (emit "drop-shadow" (String.concat " " shadows))
            in
            match c.cvalue with
            | None -> (
                match Theme.resolve_value theme None [ "--drop-shadow" ] with Some v -> wrap v | None -> not_handled)
            | Some { vkind = Arbitrary; value; _ } -> (
                if arbitrary_type c [ "color" ] = "color" then
                  match as_color value c.cmodifier theme with Some r -> color r | None -> not_handled
                else wrap value)
            | Some { value; _ } -> (
                match resolve_theme_color c [ "--drop-shadow-color"; "--color" ] theme with
                | Some theme_color -> color theme_color
                | None -> (
                    match Theme.resolve_value theme (Some value) [ "--drop-shadow" ] with
                    | Some v -> wrap v
                    | None -> not_handled)))
      end)
    [ false; true ]

File: impl/go/filters.go
package twill

import "strings"

// Filter and backdrop-filter utilities (SPEC §10.8, "Filters").

var filterNames = []string{"blur", "brightness", "contrast", "grayscale", "hue-rotate", "invert", "saturate", "sepia", "drop-shadow"}
var backdropNames = []string{"blur", "brightness", "contrast", "grayscale", "hue-rotate", "invert", "opacity", "saturate", "sepia"}

func composedFilter(prefix string, names []string) string {
	parts := make([]string, len(names))
	for i, n := range names {
		parts[i] = "var(" + prefix + n + ",)"
	}
	return strings.Join(parts, " ")
}

// Filter is the composed filter value.
var Filter = composedFilter("--tw-", filterNames)

// BackdropFilter is the composed backdrop-filter value.
var BackdropFilter = composedFilter("--tw-backdrop-", backdropNames)

func filterRegistrations(backdrop bool) []Node {
	prefix, names := "--tw-", filterNames
	if backdrop {
		prefix, names = "--tw-backdrop-", backdropNames
	}
	nodes := make([]Node, 0, len(names))
	for _, n := range names {
		nodes = append(nodes, Property(prefix+n, nil))
	}
	return nodes
}

func filterDeclarations(backdrop bool) []Node {
	if backdrop {
		return []Node{Decl("-webkit-backdrop-filter", BackdropFilter), Decl("backdrop-filter", BackdropFilter)}
	}
	return []Node{Decl("filter", Filter)}
}

// RegisterFilterUtilities registers the filter and backdrop-filter utilities.
func RegisterFilterUtilities(u *Utilities, theme *Theme) {
	statFn := func(name string, fn func() []Node) { StaticUtilityFn(u, name, fn) }
	fn := func(root string, desc FunctionalUtilityDescription) { FunctionalUtility(u, theme, root, desc) }
	percentage := func(v *CandidateValue) (string, bool) { return v.Value + "%", IsPositiveInteger(v.Value) }
	degrees := func(v *CandidateValue) (string, bool) { return v.Value + "deg", IsPositiveInteger(v.Value) }

	for _, backdrop := range []bool{false, true} {
		backdrop := backdrop
		prefix, variable := "", "--tw-"
		if backdrop {
			prefix, variable = "backdrop-", "--tw-backdrop-"
		}
		keys := func(name string) []string {
			if backdrop {
				return []string{"--backdrop-" + name, "--" + name}
			}
			return []string{"--" + name}
		}
		emit := func(name, value string) []Node {
			return append(append(filterRegistrations(backdrop), Decl(variable+name, value)), filterDeclarations(backdrop)...)
		}
		direct := func(value string) []Node {
			if backdrop {
				return []Node{Decl("-webkit-backdrop-filter", value), Decl("backdrop-filter", value)}
			}
			return []Node{Decl("filter", value)}
		}

		statFn(prefix+"filter-none", func() []Node { return direct("none") })
		statFn(prefix+"filter", func() []Node { return append(filterRegistrations(backdrop), filterDeclarations(backdrop)...) })
		fn(prefix+"filter", FunctionalUtilityDescription{Handle: func(value string, _ *string) []Node { return direct(value) }})

		fn(prefix+"blur", FunctionalUtilityDescription{ThemeKeys: keys("blur"), Handle: func(value string, _ *string) []Node { return emit("blur", "blur("+value+")") }})
		statFn(prefix+"blur-none", func() []Node { return emit("blur", "") })
		for _, name := range []string{"brightness", "contrast", "saturate"} {
			name := name
			fn(prefix+name, FunctionalUtilityDescription{ThemeKeys: keys(name), HandleBareValue: percentage,
				Handle: func(value string, _ *string) []Node { return emit(name, name+"("+value+")") }})
		}
		for _, name := range []string{"grayscale", "invert", "sepia"} {
			name := name
			fn(prefix+name, FunctionalUtilityDescription{ThemeKeys: keys(name), DefaultValue: strPtr("100%"), HandleBareValue: percentage,
				Handle: func(value string, _ *string) []Node { return emit(name, name+"("+value+")") }})
		}
		fn(prefix+"hue-rotate", FunctionalUtilityDescription{ThemeKeys: keys("hue-rotate"), SupportsNegative: true, HandleBareValue: degrees,
			Handle: func(value string, _ *string) []Node { return emit("hue-rotate", "hue-rotate("+value+")") }})
		if backdrop {
			fn("backdrop-opacity", FunctionalUtilityDescription{ThemeKeys: []string{"--backdrop-opacity", "--opacity"}, HandleBareValue: percentage,
				Handle: func(value string, _ *string) []Node { return emit("opacity", "opacity("+value+")") }})
			continue
		}

		statFn("drop-shadow-none", func() []Node { return emit("drop-shadow", "") })
		u.Functional("drop-shadow", func(c *Candidate) ([]Node, CompileStatus) {
			color := func(value string) ([]Node, CompileStatus) {
				return []Node{Property("--tw-drop-shadow-color", nil), Decl("--tw-drop-shadow-color", value)}, Handled
			}
			wrap := func(value string) ([]Node, CompileStatus) {
				failed := false
				replaced := ReplaceShadowColors(value, func(col string) string {
					resolved, ok := AsColor(col, c.Modifier, theme)
					if !ok {
						failed = true
						return col
					}
					return "var(--tw-drop-shadow-color, " + resolved + ")"
				}, false)
				if failed {
					return nil, NotHandled
				}
				var shadows []string
				for _, s := range Segment(replaced, ',') {
					shadows = append(shadows, "drop-shadow("+strings.TrimSpace(s)+")")
				}
				return emit("drop-shadow", strings.Join(shadows, " ")), Handled
			}
			if c.Value == nil {
				value, ok := theme.ResolveValue(nil, []string{"--drop-shadow"})
				if !ok {
					return nil, NotHandled
				}
				return wrap(value)
			}
			if c.Value.Kind == ValueArbitrary {
				value := c.Value.Value
				t := ""
				if c.Value.DataType != nil {
					t = *c.Value.DataType
				} else {
					t = InferDataType(value, []string{"color"})
				}
				if t == "color" {
					resolved, ok := AsColor(value, c.Modifier, theme)
					if !ok {
						return nil, NotHandled
					}
					return color(resolved)
				}
				return wrap(value)
			}
			if themeColor, ok := ResolveThemeColor(c, []string{"--drop-shadow-color", "--color"}, theme); ok {
				return color(themeColor)
			}
			value, ok := theme.ResolveValue(strPtr(c.Value.Value), []string{"--drop-shadow"})
			if !ok {
				return nil, NotHandled
			}
			return wrap(value)
		}, nil)
	}
}
